File: ra-idp/src/authentication/usecases/demo-header-resolver.ts
import {
  AuthenticationContextError,
  type AuthenticationContext,
  type AuthenticationContextResolver,
} from '../domain/authentication-context'
import type { UserRepository } from '../ports/user-repository'
import { deriveAcr } from './acr-vocabulary'

/**
 * デモ用途で X-User-Sub / X-User-Auth-Time ヘッダから AuthenticationContext を
 * 復元する。`./demo.sh` や統合テストでログイン UI を経由せずに認可フローを通すための導線。
 * 本番経路（Cookie セッション）に混入させないため LoginSessionManager とは分離する。
 */
export class DemoHeaderResolver implements AuthenticationContextResolver {
  constructor(private readonly userRepo: UserRepository) {}

  async resolve(headers: Headers): Promise<AuthenticationContext | null> {
    const sub = headers.get('x-user-sub') ?? undefined
    if (!sub) return null

    const user = await this.userRepo.findBySub(sub)
    if (!user) return null

    const amr = ['demo_header']
    return {
      sub: user.sub,
      auth_time: parseAuthTimeHeader(headers.get('x-user-auth-time') ?? undefined),
      amr,
      acr: deriveAcr(amr),
      session_id: 'header-session',
    }
  }
}

function parseAuthTimeHeader(value: string | undefined): number {
  if (!value) return Math.floor(Date.now() / 1000)
  const seconds = Number(value)
  if (!Number.isInteger(seconds) || seconds < 0) {
    throw new AuthenticationContextError('X-User-Auth-Time は Unix epoch 秒で指定してください')
  }
  return seconds
}

File: ra-idp-go/internal/authentication/usecases/demo_header_resolver.go
// DemoHeaderResolver: テスト用に "X-Demo-Sub" ヘッダから authn context を作る。
// TS src/authentication/usecases/demo-header-resolver.ts に対応。
package usecases

import (
	"context"
	"time"

	"ra-idp-go/internal/authentication/domain"
)

type DemoHeaderResolver struct{}

func (DemoHeaderResolver) Resolve(_ context.Context, h domain.Headers) (*domain.AuthenticationContext, error) {
	sub := h.Get("X-Demo-Sub")
	if sub == "" {
		return nil, nil
	}
	return &domain.AuthenticationContext{
		Sub:      sub,
		AuthTime: time.Now().Unix(),
		AMR:      []string{"pwd"},
	}, nil
}
